atomic_reactor/plugins/post_koji_upload.py

```diff
@@ -0,0 +1,639 @@
+        """
+        name = output.metadata['filename']
+        self.log.debug("uploading %r to %r as %r",
+                       output.file.name, serverdir, name)
+
+        kwargs = {}
+        if self.blocksize is not None:
+            kwargs['blocksize'] = self.blocksize
+            self.log.debug("using blocksize %d", self.blocksize)
+
+        upload_logger = KojiUploadLogger(self.log)
+        session.uploadWrapper(output.file.name, serverdir, name=name,
+                              callback=upload_logger.callback, **kwargs)
+        path = os.path.join(serverdir, name)
+        self.log.debug("uploaded %r", path)
+        return path
+
+    @staticmethod
+    def get_upload_server_dir():
+        """
+        Create a path name for uploading files to
+
+        :return: str, path name expected to be unique
+        """
+        dir_prefix = 'koji-upload'
+        random_chars = ''.join([random.choice(ascii_letters)
+                                for _ in range(8)])
+        unique_fragment = '%r.%s' % (time.time(), random_chars)
+        return os.path.join(dir_prefix, unique_fragment)
+
+    def login(self):
+        """
+        Log in to koji
+
+        :return: koji.ClientSession instance, logged in
+        """
+
+        # krbV python library throws an error if these are unicode
+        auth_info = {
+            "proxyuser": self.koji_proxy_user,
+            "ssl_certs_dir": self.koji_ssl_certs_dir,
+            "krb_principal": str(self.koji_principal),
+            "krb_keytab": str(self.koji_keytab)
+        }
+        return create_koji_session(str(self.kojihub), auth_info)
+
+    def run(self):
+        """
+        Run the plugin.
+        """
+
+        if ((self.koji_principal and not self.koji_keytab) or
+                (self.koji_keytab and not self.koji_principal)):
+            raise RuntimeError("specify both koji_principal and koji_keytab "
+                               "or neither")
+
+        # Only run if the build was successful
+        if self.workflow.build_process_failed:
+            self.log.info("Not promoting failed build to koji")
+            return
+
+        koji_metadata, output_files = self.get_metadata()
+
+        try:
+            session = self.login()
+            server_dir = self.get_upload_server_dir()
+            for output in output_files:
+                if output.file:
+                    self.upload_file(session, output, server_dir)
+        finally:
+            for output in output_files:
+                if output.file:
+                    output.file.close()
+
+        md_fragment = "{}-md".format(get_build_json()['metadata']['name'])
+        md_fragment_key = 'metadata.json'
+        cm_data = {md_fragment_key: koji_metadata}
+        annotations = {
+            "metadata_fragment": "configmap/" + md_fragment,
+            "metadata_fragment_key": md_fragment_key
+        }
+
+        try:
+            self.osbs.create_config_map(md_fragment, cm_data)
```

Comment:
My concern with serializing the data as a json string from this plugin is that consumers of this value, must also know that the data must be deserialized. This might be ok, just not a very clean API IMO.